### PR TITLE
Shorten and modernize README

### LINKED ISSUE #2746: README is too long
### Brief summary

The title says it all. The current README takes an essayistic approach. While it certainly packs in much info, it probably overwhelms the reader.

The README is one of the major entry points into an application. It should be short and inviting. Besides that, there's a good chance the info in this README is out of date (who would know? Has anyone ever read the whole thing?).

Though the format is different, the principles of what makes a good doc stay the same: be brief, be correct, and provide sufficient explanation, demonstration, and reference. Hopefully our docs do a good job with most of this, so we can be short.
Off the cuff, I'd say a new README could probably be structured something this:

- A brief statement of what k6 does and the problems it solves
- Show
  - Super short install commands
  - A Quick demo
- A few words about the overall design philosophy
- Links to major sections of the docs
- Info about contributing
- License and acknowledgments 

## Guidance

- [Art of README checklist](https://github.com/hackergrrl/art-of-readme#bonus-the-readme-checklist)
- [The README standard](https://github.com/RichardLitt/standard-readme) (Not officially stamped  by POSIX, ANSI, ISO, or IEEE, I'm afraid)
- [How to write a great README](https://dbader.org/blog/write-a-great-readme-for-your-github-project)

### k6 version

v.040

### OS

Slackware

### Docker version and image (if applicable)

_No response_

### Steps to reproduce the problem

1. Open README.md 
2. Try to read it

### Expected behaviour

It's short.

### Actual behaviour

It's long.

### PR DESCRIPTION
**Pull request by MattDodsonEnglish**
fixes #2746

### CODE REVIEW

**Comment by MattDodsonEnglish** [comment]
This is a DRAFT. I need to proofread and check links before submitting for review. But the content is unlikely to change much, so feel free to comment about something that feels over-represented or missing.

File: README.md [hunk 1]
```diff
@@ -23,350 +23,46 @@
...
 <br/>
 
-**k6** is a modern load testing tool, building on [our](https://k6.io/about) years of experience in the load and performance testing industry. It provides a clean, approachable scripting API, [local](https://k6.io/docs/getting-started/running-k6) and [cloud execution](https://k6.io/docs/cloud), and flexible configuration.
+**k6** is a modern load-testing tool, building on [our](https://k6.io/about) years of experience in the load and performance testing industries.
+k6 is built to be simple, powerful, and extensible. Every commit to this repo was intended to create an application with **high performance and excellent developer experience**.
+Some key features include:
 
-This is how load testing should look in the 21st century.
+- **Configuragle load generation.** Even lower-end machines can simulate lots of traffic.
+- **Tests as code.** Reuse scripts, share them, stick them in your CI, modularize, and version control.
```
Comment (MattDodsonEnglish) [inline README.md:31]:
```suggestion
- **Tests as code.** Reuse scripts, modularize logic, version control, and integrate tests with your CI.
```

File: README.md [hunk 2]
```diff
@@ -23,350 +23,46 @@
...
+- **A full-featured API.** The scripting API is packed with features designed to simulate real application traffic.
+- **An embedded JavaScript engine.** The performance of Go, the scripting simplicity of vanilla ECMAscript (without baggage from the wider JS ecosystem).
+- **Multiple Protocol support**. HTTP, WebSockets, gRPC, and more.
+- **Large extension ecosystem**. You can extend k6 to support your needs. Many already have and shared their work with the community!
```
Comment (MattDodsonEnglish) [inline README.md:35]:
```suggestion
- **Large extension ecosystem**. You can extend k6 to support your needs. Many already have done so and shared their extensions with the community!
```

**Comment by MattDodsonEnglish** [comment]
> For simplicity, unlike many other JavaScript runtimes, a lot of the operations in k6 are synchronous. That means that, for example, the `let response = http.get("https://test-api.k6.io/")` call from the [Running k6](#running-k6) example script will block the VU execution until the HTTP request is completed, save the [response information](https://k6.io/docs/javascript-api/k6-http/response-k6-http) in the `response` variable and only then continue executing the rest of the script - no callbacks and promises needed.

> This simplification works because k6 isn't just a single JavaScript runtime. Instead, each VU independently executes the supplied script in its own separate and semi-isolated JavaScript runtime, in parallel to all the other running VUs. This allows us to fully utilize modern multi-core hardware while at the same time lowering the script complexity by having mostly synchronous functions. Where it makes sense, we also have in-VU parallelization as well, for example, the [`http.batch()`](https://k6.io/docs/javascript-api/k6-http/batch-requests) function (which allows a single VU to make multiple simultaneous HTTP requests like a browser/real user would) or the [websocket](https://k6.io/docs/javascript-api/k6-ws) support.

I've removed this, but it actually might be useful information somewhere.

File: README.md [hunk 3]
```diff
@@ -23,350 +23,46 @@
 <img src="assets/github-hr.png" alt="---" />
 <br/>
 
-**k6** is a modern load testing tool, building on [our](https://k6.io/about) years of experience in the load and performance testing industry. It provides a clean, approachable scripting API, [local](https://k6.io/docs/getting-started/running-k6) and [cloud execution](https://k6.io/docs/cloud), and flexible configuration.
+**k6** is a modern load-testing tool, building on [our](https://k6.io/about) years of experience in the load and performance testing industries.
```
Comment (MattDodsonEnglish) [inline README.md:26]:
```suggestion
**k6** is a modern load-testing tool, building on [our years of experience](https://k6.io/about) in the load and performance testing industries.
```

File: README.md [hunk 4]
```diff
@@ -23,350 +23,46 @@
...
+k6 is built to be simple, powerful, and extensible. Every commit to this repo was intended to create an application with **high performance and excellent developer experience**.
+Some key features include:
 
-This is how load testing should look in the 21st century.
+- **Configurable load generation.** Even lower-end machines can simulate lots of traffic.
+- **Tests as code.** Reuse scripts, modularize logic, version control, and integrate tests with your CI.
+- **A full-featured API.** The scripting API is packed with features designed to simulate real application traffic.
```
Comment (MattDodsonEnglish) [inline README.md:32]:
```suggestion
- **A full-featured API.** The scripting API is packed with features that help you simulate real application traffic.
```

File: README.md [hunk 5]
```diff
@@ -23,350 +23,46 @@
...
+- **An embedded JavaScript engine.** The performance of Go, the scripting simplicity of vanilla ECMAscript (without baggage from the wider JS ecosystem).
+- **Multiple Protocol support**. HTTP, WebSockets, gRPC, and more.
+- **Large extension ecosystem**. You can extend k6 to support your needs. Many already have done so and shared their extensions with the community!
```
Comment (MattDodsonEnglish) [inline README.md:35]:
```suggestion
- **Large extension ecosystem**. You can extend k6 to support your needs. And many people have already shared their extensions with the community!
```

File: README.md [hunk 6]
```diff
@@ -23,350 +23,46 @@
...
 
-This simplification works because k6 isn't just a single JavaScript runtime. Instead, each VU independently executes the supplied script in its own separate and semi-isolated JavaScript runtime, in parallel to all the other running VUs. This allows us to fully utilize modern multi-core hardware while at the same time lowering the script complexity by having mostly synchronous functions. Where it makes sense, we also have in-VU parallelization as well, for example, the [`http.batch()`](https://k6.io/docs/javascript-api/k6-http/batch-requests) function (which allows a single VU to make multiple simultaneous HTTP requests like a browser/real user would) or the [websocket](https://k6.io/docs/javascript-api/k6-ws) support.
+The docs have much more! And if you want more conceptual information, you can read about [Test types](https://k6.io/docs/test-types/introduction/), [Test strategies](https://k6.io/docs/testing-guides/api-load-testing/), or one of the many informative [Blog posts](https://k6.io/blog)
```
Comment (MattDodsonEnglish) [inline README.md:56]:
```suggestion
These links barely scratch the surface! If you're looking for conceptual information, you can read about [Test types](https://k6.io/docs/test-types/introduction/), [Test strategies](https://k6.io/docs/testing-guides/), or one of the many informative [Blog posts](https://k6.io/blog)
```

File: README.md [hunk 7]
```diff
@@ -23,350 +23,46 @@
...
-
-You can, of course, also write your own ES6 modules and `import` them in your scripts, potentially reusing code across an organization. The situation with importing JavaScript libraries is a bit more complicated. You can potentially use **some** JS libraries in k6, even ones intended for Node.js if you use browserify, though if they depend on network/OS-related APIs, they likely won't work. You can find more details and instructions about writing or importing JS modules [here](https://k6.io/docs/using-k6/modules).
-
-
-Support
--------
+## Support
 
 To get help about usage, report bugs, suggest features and discuss k6 with other users see [SUPPORT.md](SUPPORT.md).
```
Comment (MattDodsonEnglish) [inline README.md:64]:
```suggestion
To get help about, report bugs, suggest features, and discuss k6 with other users, refer to [SUPPORT.md](SUPPORT.md).
```

File: README.md [hunk 8]
```diff
@@ -23,350 +23,46 @@
 <img src="assets/github-hr.png" alt="---" />
 <br/>
 
-**k6** is a modern load testing tool, building on [our](https://k6.io/about) years of experience in the load and performance testing industry. It provides a clean, approachable scripting API, [local](https://k6.io/docs/getting-started/running-k6) and [cloud execution](https://k6.io/docs/cloud), and flexible configuration.
+**k6** is a modern load-testing tool, building on [our years of experience](https://k6.io/about) in the load and performance testing industries.
```
Comment (MattDodsonEnglish) [inline README.md:26]:
```suggestion
**k6** is a modern load-testing tool, building on [our years of experience](https://k6.io/about) in the performance and testing industries.
```

File: README.md [hunk 9]
```diff
@@ -23,350 +23,46 @@
...
+k6 is built to be simple, powerful, and extensible. Every commit to this repo was intended to create an application with **high performance and excellent developer experience**.
```
Comment (MattDodsonEnglish) [inline README.md:27]:
```suggestion
k6 is built to be simple, powerful, and extensible. Every commit to this repo aims to help create an application with **high performance and excellent developer experience**.
```

**Comment by MattDodsonEnglish** [comment]
I also would like to change the tagline, "Like unit-testing, for performance." First, it doesn't really make sense to me. Second, I think the "like x, for y" [snowclone](https://en.wikipedia.org/wiki/Snowclone), has become pretty well-worn by now.

File: README.md [hunk 10]
```diff
@@ -23,350 +23,46 @@
...
 
-### Modules and JavaScript compatibility
+The docs cover all aspects of using k6. Some highlights include:
+- [Get Started](https://k6.io/docs). Install, run a test, inspect results.
+- [HTTP requests](https://k6.io/docs/using-k6/http-requests/). Have your virtual users use HTTP methods. Or, check the other [Protocols](https://k6.io/docs).
+- [Thresholds](https://k6.io/docs/using-k6/thresholds). Set goals for your test, and codify your SLOs.
+- [Options](https://k6.io/docs/using-k6/k6-options). Configure your load, duration, TLS certificates, and much, much more
+- [Scenarios](https://k6.io/docs/using-k6/scenarios). Choose how to model your workload: open models, closed models, constant RPS, fixed iterations, and more.
+- [Results output](https://k6.io/docs/results-output). Study, filter, and export your test results.
+- [Javascript API](https://k6.io/docs/javascript-api). Reference and examples of all k6 modules.
```
Comment (olegbespalov) [inline README.md:53]:
```suggestion
- [JavaScript API](https://k6.io/docs/javascript-api). Reference and examples of all k6 modules.
```

I think almost everywhere it has `JavaScript`, so I guess better to keep it :relaxed:

File: README.md [hunk 11]
```diff
@@ -23,350 +23,46 @@
...
-- **Built-in HAR converter**: record browser sessions as [`.har` files](https://en.wikipedia.org/wiki/.har) and [directly convert them to k6 scripts](https://k6.io/docs/using-k6/session-recording-har-support)
-- **Flexible metrics storage and visualization**: [InfluxDB](https://k6.io/docs/results-visualization/influxdb-+-grafana) (+Grafana), [JSON](https://k6.io/docs/getting-started/results-output/json) or [k6 Cloud](https://k6.io/docs/cloud/analyzing-results/overview)
-- [**Cloud execution**](https://k6.io/docs/using-k6/cloud-execution) and distributed tests _(currently only on infrastructure managed by [us k6 cloud](https://k6.io/cloud), with native distributed execution in k6 [planned](https://github.com/grafana/k6/issues/140) for the near future!)_
-
-There's even more! [See all features available in k6.](https://k6.io/docs/)
-
-Install
```
Comment (olegbespalov) [inline README.md:58]: We reefer to this section from top:
image

So it's also should be adjusted

File: README.md [hunk 12]
```diff
@@ -23,350 +23,46 @@
...
+- **A full-featured API.** The scripting API is packed with features that help you simulate real application traffic.
+- **An embedded JavaScript engine.** The performance of Go, the scripting simplicity of vanilla ECMAscript (without baggage from the wider JS ecosystem).
+- **Multiple Protocol support**. HTTP, WebSockets, gRPC, and more.
+- **Large extension ecosystem.** You can extend k6 to support your needs. And many people have already shared their extensions with the community!
+- **Flexible metrics storage and visualization**. Summary statistics or granular metrics, exported to the service of your choice.
+
+This is what load testing looks like in the 21st century.
 
 <p align="center">
   <img width="600" src="assets/k6-demo.gif">
```
Comment (mstoykov) [inline README.md:41]: This Demo is 3 years old and likely should be updated.

Also as a recommendation can we not store "assets" that aren't part of the project in the repo and instead upload them to github by just copying them in an issue and then using the generated link (no need to post the issue).

**Review by mstoykov — APPROVED** [review]
Thanks a lot for data :clap: 

I had a single comment that likely can be a separate issue.

File: README.md [hunk 13]
```diff
@@ -23,350 +23,46 @@
...
 
-You can, of course, also write your own ES6 modules and `import` them in your scripts, potentially reusing code across an organization. The situation with importing JavaScript libraries is a bit more complicated. You can potentially use **some** JS libraries in k6, even ones intended for Node.js if you use browserify, though if they depend on network/OS-related APIs, they likely won't work. You can find more details and instructions about writing or importing JS modules [here](https://k6.io/docs/using-k6/modules).
+These links barely scratch the surface! If you're looking for conceptual information, you can read about [Test types](https://k6.io/docs/test-types/introduction/), [Test strategies](https://k6.io/docs/testing-guides/), or one of the many informative [Blog posts](https://k6.io/blog)
```
Comment (codebien) [inline README.md:56]:
```suggestion
These links barely scratch the surface! If you're looking for conceptual information, you can read about [Test types](https://k6.io/docs/test-types/introduction/), [Test strategies](https://k6.io/docs/testing-guides/), or one of the many informative [Blog posts](https://k6.io/blog).
```

File: README.md [hunk 14]
```diff
@@ -23,350 +23,46 @@
...
 
+## Contribute
 
-Support
--------
+If you want to contribute or help with the development of k6, start by reading [CONTRIBUTING.md](CONTRIBUTING.md). Before you start coding, it might be a good idea to first discuss your plans and implementation details with the k6 maintainers-especially when it comes to big changes and features. You can do this either in the [GitHub issue](https://github.com/grafana/k6/issues) for the problem you're solving (create one if it doesn't exist) or on the [k6 community forum](https://community.k6.io).
```
Comment (codebien) [inline README.md:60]:
```suggestion
If you want to contribute or help with the development of k6, start by reading [CONTRIBUTING.md](CONTRIBUTING.md). Before you start coding, it might be a good idea to first discuss your plans and implementation details with the k6 maintainers-especially when it comes to big changes and features. You can do this in the [GitHub issue](https://github.com/grafana/k6/issues) for the problem you're solving (create one if it doesn't exist).
```
I think we don't expect people to discuss contributions in the forum.

**Review by codebien — COMMENTED** [review]
Thanks for doing it :heart:

**Review by oleiade — DISMISSED** [review]
Great work @MattDodsonEnglish 🎉 Thanks for taking care of it 🙇🏻 

One non-blocking comment from my side (hence, approving regardless): As a GitHub user, I wouldn't click on links to get examples of how to use something.
It's actually one of the first things I look for: an example of what it looks like to use it. If it's not there, or if I have to be redirected to a separate website to get that, I just leave without further ceremony. As a GitHub user, immediately showing me how it's used is how you get me (I got trapped in using projects which looked great on paper, but terrible in practice, in the past).
There is a big chance I landed on a README by chance because someone pointed it out to me, or someone referenced it: now that I'm there, convince me that you're easier to use than X or Y; it's basically marketing.
We can assume I'm the outlier here, but I just wanted to put it out there.

File: README.md [hunk 15]
```diff
@@ -23,350 +23,46 @@
...
+**k6** is a modern load-testing tool, building on [our years of experience](https://k6.io/about) in the performance and testing industries.
+k6 is built to be simple, powerful, and extensible. Every commit to this repo aims to help create an application with **high performance and excellent developer experience**.
+Some key features include:
 
-This is how load testing should look in the 21st century.
+- **Configurable load generation.** Even lower-end machines can simulate lots of traffic.
+- **Tests as code.** Reuse scripts, modularize logic, version control, and integrate tests with your CI.
+- **A full-featured API.** The scripting API is packed with features that help you simulate real application traffic.
+- **An embedded JavaScript engine.** The performance of Go, the scripting simplicity of vanilla ECMAscript (without baggage from the wider JS ecosystem).
```
Comment (imiric) [inline README.md:33]: Two points here:
- "vanilla ECMAScript" doesn't really say much. What exactly would "non-vanilla" ES be?
- I think "without baggage from the wider JS ecosystem" is an unnecessary swipe.

So we might want to rephrase this as:

> The performance of Go, with the familiarity of JavaScript.

File: README.md [hunk 16]
```diff
@@ -23,350 +23,46 @@
...
+- **Multiple Protocol support**. HTTP, WebSockets, gRPC, and more.
+- **Large extension ecosystem.** You can extend k6 to support your needs. And many people have already shared their extensions with the community!
```
Comment (imiric) [inline README.md:35]: Maybe let's link to the Extensions docs page here?

**Review by imiric — CHANGES_REQUESTED** [review]
Thanks @MattDodsonEnglish! :clap:

This is much cleaner now, but I think we should change that "vanilla ECMAScript" line, and I also agree with @oleiade that seeing script examples quickly would help a lot. We don't need to have them embedded in the README, but it would be good to link to the `samples/` directory, if nothing else.

File: README.md (hunk 15 above)
Comment (MattDodsonEnglish) [reply README.md:33]: Yeah you're totally right. I was under the influence of a discussion from last week :-).

Comment (MattDodsonEnglish) [inline README.md:33]:
```suggestion
- **An embedded JavaScript engine.** The performance of Go, with the familiar scripting of ECMAscript.
```

Comment (MattDodsonEnglish) [reply README.md:33]: The drawback of what I just said is that ECMAScript is an absolutely hideous word, hard on both eye and ear.

File: README.md (hunk 16 above)
Comment (MattDodsonEnglish) [reply README.md:35]: To stay consistent, I'd like to leave this section linkless, but I should add a link and bullet for extensions in the next section :+1:

File: README.md (hunk 15 above)
Comment (imiric) [reply README.md:33]: I don't think we need to mention ES, unless we're discussing the language specification in the context of the features we support. Since this is a high-level overview, I think we can stick to just JavaScript. There's nothing "vanilla" about ES; or with k6/goja's implementation of it. Repeating JS doesn't bother me.

But maybe someone else can chime in.

**Comment by MattDodsonEnglish** [comment]
I agree with Oleaide that examples are super important. The reasons I left them out here:
- The Gif. But it's old, as Mihail pointed out.
- To standardize with other Grafana READMEs. I was following the example of Grafana, Mimir, Loki.
- I thought that, since k6 is such a big project, most will look to the docs to see how things work? Speaking for my own habits, if it's some little utility tool or very specific project, I usually look on the README first. If it's some big, well-known project, like Docker or Kubernetes or Redis, I usually look at the docs first. 


That said, I think it'd be good to have a demo.

Maybe we could have a quick example section, using collapsibles. Something like:

----
## Demo

After installing, here's how you can use k6:

1. Write your script in a JavaScript file:
<details><summary>test.js</summary>
There's no real test script yet
</details>
2. Use k6 CLI to run it
  ```bash
  k6 run test.js
  ```
3. Study the a test summary, or study every data point in detail (in a file, or with an external service).
----

What y'all think?

File: README.md [hunk 17]
```diff
@@ -23,350 +23,46 @@
...
+**k6** is a modern load-testing tool, building on [our years of experience](https://k6.io/about) in the performance and testing industries.
+k6 is built to be simple, powerful, and extensible. Every commit to this repo aims to help create an application with **high performance and excellent developer experience**.
+Some key features include:
 
-This is how load testing should look in the 21st century.
+- **Configurable load generation.** Even lower-end machines can simulate lots of traffic.
+- **Tests as code.** Reuse scripts, modularize logic, version control, and integrate tests with your CI.
+- **A full-featured API.** The scripting API is packed with features that help you simulate real application traffic.
+- **An embedded JavaScript engine.** The performance of Go, the scripting simplicity of vanilla ECMAScript (without baggage from the wider JS ecosystem).
```
Comment (MattDodsonEnglish) [inline README.md:33]:
```suggestion
- **An embedded JavaScript engine.** The performance of Go, the scripting simplicity of  JavaScript.
```

**Comment by imiric** [comment]
> To standardize with other Grafana READMEs. I was following the example of Grafana, Mimir, Loki.

I think k6 is a bit different from these projects, since a JS script is part of the user experience, which can't be easily showcased for these projects. If visitors can quickly see what that looks like from the README, instead of going to an external docs site, or inspecting our `samples/` directory, I think it's a slight improvement that doesn't add too much clutter to the page (as long as it's just one simple example, not something too complex).

For example, [Locust](https://github.com/locustio/locust) is a comparable project which shows a small example in Python. But as a counter point, [Gatling](https://github.com/gatling/gatling) doesn't, and I think it hurts it, because I have no intention to dig into the docs to see what that looks like.

For demo purposes we have the animation (hopefully in a crisp MP4 soon :wink:), so we don't need to have instructions for actually using k6. But I think we should show a small example script inline, without hiding it, simply as a showcase of what a k6 script looks like.

File: README.md [hunk 18]
```diff
@@ -23,350 +23,46 @@
...
+## Contribute
 
-Support
--------
+If you want to contribute or help with the development of k6, start by reading [CONTRIBUTING.md](CONTRIBUTING.md). Before you start coding, it might be a good idea to first discuss your plans and implementation details with the k6 maintainers-especially when it comes to big changes and features. You can do this either in the [GitHub issue](https://github.com/grafana/k6/issues) for the problem you're solving (create one if it doesn't exist) or on the [k6 community forum](https://community.k6.io).
 
-To get help about usage, report bugs, suggest features and discuss k6 with other users see [SUPPORT.md](SUPPORT.md).
+## Support
 
+To get help, report bugs, suggest features, and discuss k6 with others, refer to [SUPPORT.md](SUPPORT.md).
```
[the remark this answers is not in the document]
Comment (MattDodsonEnglish) [reply README.md:96]: Good point! I added links in the readme and in the contributing file. I think it's more contributing than support, no? I'm okay with adding it to SUPPORT.md, too.

File: README.md (hunk 11 above)
Comment (MattDodsonEnglish) [reply README.md:58]: Thanks, didn't realize that at all! Removed.

File: README.md [hunk 19]
```diff
@@ -23,350 +22,80 @@
...
-
-[Checks](https://k6.io/docs/using-k6/checks) are similar to asserts but differ in that they don't halt execution. Instead, they just store the result of the check, pass or fail, and let the script execution continue. Checks are great for codifying assertions relating to HTTP requests/responses. For example, making sure an HTTP response code is 2xx.
-
-[Thresholds](https://k6.io/docs/using-k6/thresholds) are global pass/fail criteria that can be used to verify if any result metric is within a specified range. They can also reference a subset of values in a given metric, based on the used metric tags. Thresholds are specified in the options section of a k6 script. If they are exceeded during a test run, k6 would exit with a nonzero code on test completion, and can also optionally abort the test early. This makes thresholds ideally suited as checks in a CI workflow!
+Here's a basic k6 script.
+Besides simulating a user getting a resource, the test also verifies response data and response time over changing load.
```
Comment (MattDodsonEnglish) [inline README.md:42]:
```suggestion

```

File: README.md [hunk 20]
```diff
@@ -23,350 +22,78 @@
...
+  // Ramp the number virtual users up and down
+  stages: [
+    { duration: "30s", target: 15 },
+    { duration: "1m", target: 15 },
+    { duration: "20s", target: 0 },
+  ],
 };
 
-// Main function
+// code the user logic
```
Comment (MattDodsonEnglish) [inline README.md:59]:
```suggestion
// The virtual user behavior logic
```

File: README.md [hunk 21]
```diff
@@ -23,350 +22,78 @@
...
-        // Thresholds based on the custom metric we defined and use to track application failures
-        "check_failure_rate": [
-            // Global failure rate should be less than 1%
-            "rate<0.01",
-            // Abort the test early if it climbs over 5%
-            { threshold: "rate<=0.05", abortOnFail: true },
-        ],
-    },
+import { check, sleep } from "k6";
+
```
Comment (MattDodsonEnglish) [inline README.md:45]:
```suggestion

// test configuration
```

File: README.md [hunk 22]
```diff
@@ -23,350 +22,80 @@
...
+The docs cover all aspects of using k6. Some highlights include:
+- [Get Started](https://k6.io/docs). Install, run a test, inspect results.
+- [HTTP requests](https://k6.io/docs/using-k6/http-requests/). Have your virtual users use HTTP methods.
+  Or, check the other [Protocols](https://k6.io/docs).
+- [Thresholds](https://k6.io/docs/using-k6/thresholds). Set goals for your test, and codify your SLOs.
+- [Options](https://k6.io/docs/using-k6/k6-options). Configure your load, duration, TLS certificates, and much, much more
```
Comment (imiric) [inline README.md:79]:
```suggestion
- [Options](https://k6.io/docs/using-k6/k6-options). Configure your load, duration, TLS certificates, and much, much more.
```

File: README.md [hunk 23]
```diff
@@ -23,350 +22,80 @@
...
+## Contribute
 
-You can, of course, also write your own ES6 modules and `import` them in your scripts, potentially reusing code across an organization. The situation with importing JavaScript libraries is a bit more complicated. You can potentially use **some** JS libraries in k6, even ones intended for Node.js if you use browserify, though if they depend on network/OS-related APIs, they likely won't work. You can find more details and instructions about writing or importing JS modules [here](https://k6.io/docs/using-k6/modules).
+If you want to contribute or help with the development of k6, start by reading [CONTRIBUTING.md](CONTRIBUTING.md). Before you start coding, it might be a good idea to first discuss your plans and implementation details with the k6 maintainers-especially when it comes to big changes and features. You can do this in the [GitHub issue](https://github.com/grafana/k6/issues) for the problem you're solving (create one if it doesn't exist).
```
Comment (imiric) [inline README.md:90]:
```suggestion
If you want to contribute or help with the development of k6, start by reading [CONTRIBUTING.md](CONTRIBUTING.md). Before you start coding, it might be a good idea to first discuss your plans and implementation details with the k6 maintainers—especially when it comes to big changes and features. You can do this in the [GitHub issue](https://github.com/grafana/k6/issues) for the problem you're solving (create one if it doesn't exist).
```

File: README.md [hunk 24]
```diff
@@ -23,350 +22,80 @@
...
-And for rpm-based ones like Fedora and CentOS:
-
-```bash
-sudo dnf install https://dl.k6.io/rpm/repo.rpm    # use yum instead of dnf for older distros
-sudo dnf install k6    # use yum install --nogpgcheck k6 for older distros (e.g. CentOS 7, Amazon Linux 2) without PGP V4 signature support
-```
+- **Configurable load generation.** Even lower-end machines can simulate lots of traffic.
+- **Tests as code.** Reuse scripts, modularize logic, version control, and integrate tests with your CI.
+- **A full-featured API.** The scripting API is packed with features that help you simulate real application traffic.
+- **An embedded JavaScript engine.** The performance of Go, the scripting simplicity of JavaScript.
```
Comment (imiric) [inline README.md:33]: Nit: I wouldn't say the benefits of JS are in its simplicity (Go is arguably a simpler language). The decision to script in JS was mostly because of its widespread use and familiarity.

**Review by imiric — APPROVED** [review]
Great work @MattDodsonEnglish! :clap:

Just left some tiny nitpicks, feel free to disregard.

File: README.md [hunk 25]
```diff
@@ -23,350 +22,80 @@
 <img src="assets/github-hr.png" alt="---" />
 <br/>
 
-**k6** is a modern load testing tool, building on [our](https://k6.io/about) years of experience in the load and performance testing industry. It provides a clean, approachable scripting API, [local](https://k6.io/docs/getting-started/running-k6) and [cloud execution](https://k6.io/docs/cloud), and flexible configuration.
+**k6** is a modern load-testing tool, built on [our years of experience](https://k6.io/about) in the performance and testing industries.
+k6 is built to be powerful, extensible, and full-featured. The key design goals are **high performance and the best developer experience**.
```
Comment (codebien) [inline README.md:26]:
```suggestion
k6 is built to be powerful, extensible, and full-featured. The key design goals are **the best developer experience** and high performance.
```
Based on the last company-wide discussion and the different perceptions of performance we got I guess we should at least reverse the order or reduce the emphasis on high performance. If we want to have it (I can understand that) then maybe should we explain it a bit more?

File: README.md (hunk 24 above)
Comment (MattDodsonEnglish) [inline README.md:33]:
```suggestion
- **An embedded JavaScript engine.** The performance of Go, the scripting familiarity of JavaScript.
```

File: README.md (hunk 25 above)
Comment (MattDodsonEnglish) [reply README.md]: This is a good point. I've just removed the "high performance."

This brings up a huge topic: What are the key design goals of k6? What gets sacrificed when a tradeoff has been made between performance and ease of use? Why implement libraries instead of having testers write their own helper functions.

These questions probably require participation from a large amount of stakeholders.
But, this discussion is too important to have in a side conversation about a Readme fix. I'm going to make a side issue.

File: README.md [hunk 26]
```diff
@@ -23,350 +22,80 @@
 <img src="assets/github-hr.png" alt="---" />
 <br/>
 
-**k6** is a modern load testing tool, building on [our](https://k6.io/about) years of experience in the load and performance testing industry. It provides a clean, approachable scripting API, [local](https://k6.io/docs/getting-started/running-k6) and [cloud execution](https://k6.io/docs/cloud), and flexible configuration.
+**k6** is a modern load-testing tool, built on [our years of experience](https://k6.io/about) in the performance and testing industries.
+k6 is built to be powerful, extensible, and full-featured. The key design goals is to provide **the best developer experience**.
```
Comment (imiric) [inline README.md:26]:
```suggestion
It's built to be powerful, extensible, and full-featured. The key design goal is to provide **the best developer experience**.
```

Currently we repeat "k6 ..." too much.

File: README.md [hunk 27]
```diff
@@ -23,350 +22,80 @@
...
 
-This is how load testing should look in the 21st century.
+k6 comes with the batteries included:
```
Comment (imiric) [inline README.md:28]:
```suggestion
Its core features are:
```

**Review by imiric — DISMISSED** [review]

File: README.md (hunk 27 above)
Comment (MattDodsonEnglish) [reply README.md:28]: I'm going to keep this, because "batteries included" is as of now a central design consideration of the application. I actually originally took it out, because it's jargony, but in a separate conversation with @sniku, I realized it's an important part of what k6 tries to be, so I put it back. Still I'd like a less metaphorical description, but let's let it stand for now :-).

Comment (imiric) [reply README.md:28]: I do think it's jargony, as you say, but OK. In that case, I think we can drop "the", so "k6 comes with batteries included:".

Comment (MattDodsonEnglish) [reply README.md:28]: Ah, on second thought I got "writer's remorse" for using a phrase that's metaphorical. Batteries included makes it sound like it could be a CD player, and k6 is not a CD player. I think "full-featured", in the previous paragraph, expresses what we want to say with "batteries included".

**Review by imiric — DISMISSED** [review]

**Review by codebien — APPROVED** [review]
LGTM, just two nitpick

**Review by imiric — DISMISSED** [review]
This is like my 5th time approving this :sweat_smile:

Can we get this merged soon? :pray:

**Review by oleiade — APPROVED** [review]
LGTM 👏🏻 🎉 🦖 Thanks for that @MattDodsonEnglish great work!

File: README.md [hunk 28]
```diff
@@ -23,350 +22,80 @@
...
-Scripts must contain, at the very least, an exported `default` function - this defines the entry point for your VUs, similar to the `main()` function in many languages. Let's create a very simple script that makes an HTTP GET request to a test website:
 
 ```js
 import http from "k6/http";
-
-export default function() {
-    let response = http.get("https://test-api.k6.io");
+import { check, sleep } from "k6";
+
+// test configuration
```
[the remark this answers is not in the document]
Comment (codebien) [reply README.md:47]: @MattDodsonEnglish do you plan to address this or not? If not, let me know so I can approve.

**Review by olegbespalov — APPROVED** [review]
:tada: :muscle:

**Review by codebien — APPROVED** [review]
:tada: Thanks @MattDodsonEnglish

**Review by imiric — APPROVED** [review]
